swarm/api/manifest.go

```diff
@@ -19,34 +19,167 @@ package api
 import (
 	"bytes"
 	"encoding/json"
+	"errors"
 	"fmt"
+	"io"
 	"sync"
+	"time"
 
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/log"
 	"github.com/ethereum/go-ethereum/swarm/storage"
 )
 
 const (
-	manifestType = "application/bzz-manifest+json"
+	ManifestType = "application/bzz-manifest+json"
 )
 
+// Manifest represents a swarm manifest
+type Manifest struct {
+	Entries []ManifestEntry `json:"entries,omitempty"`
+}
+
+// ManifestEntry represents an entry in a swarm manifest
+type ManifestEntry struct {
```

Comment:
> This looks like a breaking change. If you dont have Path here old manifests will not parse with this?

`ManifestEntry` is embedding `FileInfo` which means it inherits all the fields, so `ManifestEntry.Path` still exists here and will serialize to JSON in the same way, so it will not break old manifests.

> We must put the path outside FileInfo anyway. Path is not a file info in the manifest context, it is a routing instruction ... Same goes for content_type.

What is wrong with having `Path` on `FileInfo`? This makes it very easy for clients to add files with the attributes they want, for example:

* add a file which I can access using the path `some/path`, content type `text/plain` and permissions `0644`:

```go
client.Upload(&FileInfo{
	Path: "some/path",
	ContentType: "text/plain",
	Mode: 0644,
})
```

If this is confusing (perhaps you are thinking of `os.FileInfo`) then perhaps we should choose a different name? Perhaps just drop `FileInfo` and use `ManifestEntry` directly in the clients?